runtime/common/src/main/kotlin/kotlinx/serialization/context/Context.kt

```diff
@@ -41,7 +57,45 @@ fun <T: Any> SerialContext?.getOrDefault(klass: KClass<T>) = this?.let { get(kla
 @ImplicitReflectionSerializer
 fun <T: Any> SerialContext?.getByValueOrDefault(value: T): KSerializer<T> = this?.let { getByValue(value) } ?: value::class.serializer() as KSerializer<T>
 
+/**
+ * A [SerialContext] which always returns `null`.
+ */
 object EmptyContext: SerialContext {
     override fun <T : Any> get(kclass: KClass<T>): KSerializer<T>? = null
-    override fun <T : Any> getByValue(value: T): KSerializer<T>? = null
+    override fun <T : Any> resolveFromBase(basePolyType: KClass<T>, obj: T): KSerializer<out T>? =
+        null
+
+    override fun <T : Any> resolveFromBase(basePolyType: KClass<T>, serializedClassName: String): KSerializer<out T>? =
+        null
+}
+
+/**
+ * A [SerialContext] which has ability to register a [KSerializer] in it.
+ *
+ * @see SerialModule
+ */
+interface MutableSerialContext: SerialContext {
+    /**
+     * Registers a [serializer] for a [forClass].
+     *
+     * Given [serializer] will be used in [get] and [getByValue].
+     */
+    fun <T: Any> registerSerializer(forClass: KClass<T>, serializer: KSerializer<T>)
+
+    /**
+     * Registers a [concreteSerializer] for a [concreteClass] in a scope of [basePolyType].
+     *
+     * Given [concreteSerializer] will be used in polymorphic serialization
+     * if statically known type of property is [basePolyType] and runtime type is a [concreteClass].
+     *
+     * For serialization process, classname is determined using name of a serializer descriptor.
+     *
+     * All standard types and collections are registered in advance as a subtypes in a scope of [kotlin.Any].
+     */
+    fun <Base : Any, Sub : Base> registerPolymorphicSerializer(
```

Comment:
also, it is a good idea to throw an exception if users try to register multiple serializers for the same class

Reply:
A good default will be to throw an exception by default but provide an additional argument `overwriteExisting` (with proper naming)  and note this parameter in the message of the exception.

WDYT?

Reply:
Sounds ok, but should also add it to usual `registerSerializer` then

Reply:
And since clients of these functions are modules, we should somehow propagate this parameter to all default modules (`SingletonModule`, `MapModule`, `PolymorphicModule`, etc). I wonder how far this can go. Probably we should leave this to design then.